core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -77,19 +78,24 @@ protected TransportWriteAction(Settings settings, String actionName, TransportSe
      * async refresh is performed on the <code>replica</code> shard according to the <code>ReplicaRequest</code> refresh policy
      */
     @Override
-    protected abstract WriteReplicaResult shardOperationOnReplica(ReplicaRequest request, IndexShard replica) throws Exception;
+    protected abstract WriteReplicaResult<ReplicaRequest> shardOperationOnReplica(
+            ReplicaRequest request, IndexShard replica) throws Exception;
 
     /**
      * Result of taking the action on the primary.
      */
-    protected class WritePrimaryResult extends PrimaryResult implements RespondingWriteResult {
+    protected static class WritePrimaryResult<ReplicaRequest extends ReplicatedWriteRequest<ReplicaRequest>,
```

Comment:
this is needed because we deal with two types of `WritePrimaryResult` in `indexAction/deleteAction.shardOperationOnPrimary/Replica`. `WritePrimaryResult` with type `BulkShardRequest/BulkShardResponse` and `Index`/`delete` request and response.